usr/src/uts/common/fs/zfs/zfs_acl.c

```diff
@@ -1716,15 +1731,18 @@ zfs_acl_ids_create(znode_t *dzp, int flag, vattr_t *vap, cred_t *cr,
 		mutex_exit(&dzp->z_lock);
 		mutex_exit(&dzp->z_acl_lock);
 
-		if (vap->va_type == VDIR)
-			acl_ids->z_aclp->z_hints |= ZFS_ACL_AUTO_INHERIT;
-
-		if (zfsvfs->z_acl_mode == ZFS_ACL_GROUPMASK &&
-		    zfsvfs->z_acl_inherit != ZFS_ACL_PASSTHROUGH &&
-		    zfsvfs->z_acl_inherit != ZFS_ACL_PASSTHROUGH_X)
-			trim = B_TRUE;
-		zfs_acl_chmod(vap->va_type, acl_ids->z_mode, B_FALSE, trim,
-		    acl_ids->z_aclp);
+		if (need_chmod) {
```

Comment:
How would you feel about adding a tuneable to make this unconditional?  Something like zfs_acl_always_chmod maybe.

Reply:
For what purpose? need_chmod defaults to true, and it's only set to false if aclinherit is passthrough or passthrough-x and there are trivial acl entries being inherited (to match the documented behavior of these modes). Setting that proposed tuning variable would just result in the broken ACL inheritance that this patch fixes.

If you want a behavior like "inherit everything but then chmod anyway" it seems that should be a new type of aclinherit?

Reply:
I agree with @pbhenson here. This change is about reverting behaviour that changed in 6764 so that it again works as documented (and expected).

The chmod step is only skipped if there are inheritable ACLs that directly affect the file mode (i.e. user@, group@, everyone@) and that makes sense since the reason to set these is to override the standard `ugo` permissions. Other inheritable ACLs (e.g. user: or group:) don't suppress the chmod.

I can't see a scenario where unconditional chmod would be required but it does seem like another value for `aclinherit` and not part of this change.

Reply:
I think part of the problem might be that there are those already relying on supposedly broken functionality. A tunable would allow for a softer landing, at least in some cases. This will certainly be true for customers of RackTop, so I definitely support @andy-js suggestion.